src/gasPriceOracle/oracle.ts

```diff
@@ -0,0 +1,92 @@
+import { BigNumber, providers } from "ethers";
+import { toBN } from "../utils";
+
+export type GasPriceEstimate = {
+  maxFeePerGas: BigNumber;
+  maxPriorityFeePerGas?: BigNumber;
+};
+
+type ValidatedFeeData = {
+  gasPrice: BigNumber;
+  maxFeePerGas: BigNumber;
+  maxPriorityFeePerGas: BigNumber;
+};
+
+interface GasPriceFeed {
+  (provider: providers.Provider): Promise<GasPriceEstimate>;
+}
+
+async function getProviderFeeData(provider: providers.Provider): Promise<ValidatedFeeData> {
+  const validateFeeData = (val: BigNumber | null): BigNumber  => {
+    return BigNumber.isBigNumber(val) && val.gt(0) ? val : toBN(Number.MAX_SAFE_INTEGER);
+  };
+
+  // Not much to do on a failed retrieval, so just suppress it.
+  const response: providers.FeeData = await provider.getFeeData().catch(() => {
+    return {} as providers.FeeData
+  });
+
+  const feeData: ValidatedFeeData = {
+    "gasPrice": validateFeeData(response["gasPrice"]),
+    "maxFeePerGas": validateFeeData(response["maxFeePerGas"]),
+    "maxPriorityFeePerGas": validateFeeData(response["maxPriorityFeePerGas"]),
+  };
+
+  return feeData;
+}
+
+async function legacy(provider: providers.Provider): Promise<GasPriceEstimate> {
+  const feeData: ValidatedFeeData = await getProviderFeeData(provider);
+
+  return {
+    maxFeePerGas: feeData.gasPrice,
+  } as GasPriceEstimate;
+}
+
+// @todo: Update to ethers 5.7.x to access FeeData.lastBaseFeePerGas
+// https://github.com/ethers-io/ethers.js/commit/8314236143a300ae81c1dcc27a7a36640df22061
+async function eip1559(provider: providers.Provider): Promise<GasPriceEstimate> {
+  const feeData: ValidatedFeeData = await getProviderFeeData(provider);
+  const maxPriorityFeePerGas: BigNumber = feeData.maxPriorityFeePerGas;
+
+  return {
+    maxPriorityFeePerGas: maxPriorityFeePerGas,
+    maxFeePerGas: maxPriorityFeePerGas.add(feeData.maxFeePerGas),
+  } as GasPriceEstimate;
```

Comment:
nit: why does this need to be casted? Shouldn't it conform to the return type you specified without a cast?